x-pack/plugins/security_solution/public/detections/components/alerts_table/timeline_actions/alert_context_menu.tsx

```diff
@@ -0,0 +1,446 @@
+import { Ecs, TimelineNonEcsData } from '../../../../graphql/types';
+import {
+  AddExceptionModal as AddExceptionModalComponent,
+  AddExceptionModalBaseProps,
+} from '../../../../common/components/exceptions/add_exception_modal';
+import { getMappedNonEcsValue } from '../../../../common/components/exceptions/helpers';
+import * as i18n from '../translations';
+import {
+  useStateToaster,
+  displaySuccessToast,
+  displayErrorToast,
+} from '../../../../common/components/toasters';
+import { useUserData } from '../../user_info';
+
+interface AlertContextMenuProps {
+  ecsRowData: Ecs;
+  nonEcsRowData: TimelineNonEcsData[];
+  timelineId: string;
+}
+
+const addExceptionModalInitialState: AddExceptionModalBaseProps = {
+  ruleName: '',
+  ruleId: '',
+  ruleIndices: [],
+  exceptionListType: 'detection',
+  alertData: undefined,
+};
+
+export const AlertContextMenu = React.memo<AlertContextMenuProps>(
+  ({ ecsRowData, nonEcsRowData, timelineId }) => {
+    const dispatch = useDispatch();
+    const [, dispatchToaster] = useStateToaster();
+    const [isPopoverOpen, setPopover] = useState(false);
+    const alertStatus =
+      (ecsRowData.signal?.status && (ecsRowData.signal.status[0] as Status)) ?? undefined;
+    const eventId = ecsRowData._id;
+
+    const onButtonClick = useCallback(() => {
+      setPopover(!isPopoverOpen);
+    }, [isPopoverOpen]);
+
+    const closePopover = useCallback(() => {
+      setPopover(false);
+    }, []);
+    const [shouldShowAddExceptionModal, setShouldShowAddExceptionModal] = useState(false);
+    const [addExceptionModalState, setAddExceptionModalState] = useState<
+      AddExceptionModalBaseProps
+    >(addExceptionModalInitialState);
+    const [{ canUserCRUD, hasIndexWrite }] = useUserData();
+
+    const isEndpointAlert = useMemo(() => {
+      if (!nonEcsRowData) return false;
+      const [module] = getMappedNonEcsValue({
+        data: nonEcsRowData,
+        fieldName: 'signal.original_event.module',
+      });
+      const [kind] = getMappedNonEcsValue({
+        data: nonEcsRowData,
+        fieldName: 'signal.original_event.kind',
+      });
+      return module === 'endpoint' && kind === 'alert';
+    }, [nonEcsRowData]);
+
+    const closeAddExceptionModal = useCallback(() => {
+      setShouldShowAddExceptionModal(false);
+      setAddExceptionModalState(addExceptionModalInitialState);
+    }, [setShouldShowAddExceptionModal, setAddExceptionModalState]);
+
+    const onAddExceptionCancel = useCallback(() => {
+      closeAddExceptionModal();
+    }, [closeAddExceptionModal]);
+
+    const onAddExceptionConfirm = useCallback(() => closeAddExceptionModal(), [
+      closeAddExceptionModal,
+    ]);
+
+    const onAlertStatusUpdateSuccess = useCallback(
+      (count: number, newStatus: Status) => {
+        let title: string;
+        switch (newStatus) {
+          case 'closed':
+            title = i18n.CLOSED_ALERT_SUCCESS_TOAST(count);
```

Comment:
Per our out-of-band conversation yesterday, this PR won't, for example, refresh the `Detections` page when the status of an alert is changed in a timeline. (That functionality will come in a follow-up PR that refactors the timeline query.) That said, perhaps we can (in this PR) update the list of valid overflow menu actions here when users change the status of events?

To reproduce the current behavior:

1) Navigate to the Detections page

2) Click on any open alert's `Investigate in timeline` action

**Expected result**
* The open alert is displayed in timeline

3) Click on the alert's overflow menu

**Expected results**
- The `Mark in progress` and `Close alert` actions appear in the menu

4) Click the `Close alert` action from the overflow menu

**Expected results**
* The "Successfully closed 1 alert" toast appears

4) Once again, click on the overflow menu

**Expected results**
* The overflow menu no-longer displays the `Close alert` action
* The `Open alert` and `Mark in progress` actions appear in the menu

**Actual results**
* The `Close alert` action still appears in the menu
* The `Open alert` action does not appear in the menu

Reply:
Regarding the comment above: I think this separate from the

> **KNOWN ISSUE**: Timeline results don't refresh after the alert status change, will be fixed after https://github.com/elastic/kibana/pull/75591 is merged

noted in the description of the PR.

I think it, for example, makes sense to refresh the `Detections` page when an alert is closed in the active timeline, which will be fixed by the above PR, but I don't think the active timeline should re-execute it's query.

Re-executing the query in the active timeline when an alert is closed would, for example, reset the active timeline's scroll position.

At this point in the code, we know (at the row level) whether or not the status was successfully updated, so we could potentially perform a row-level update here that would ensure the action menu behaves as-expected, without re-running the timeline query for the active timeline.